application-diagram-api/src/main/java/com/xwiki/diagram/internal/DiagramMacroRunnable.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.diagram.internal;
+
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReferenceSerializer;
+import org.xwiki.rendering.block.Block;
+import org.xwiki.rendering.block.Block.Axes;
+import org.xwiki.rendering.block.XDOM;
+import org.xwiki.rendering.block.match.MacroBlockMatcher;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+
+/**
+ * Updates diagram macro reference parameter after the rename of a diagram.
+ * 
+ * @version $Id$
+ * @since 1.14
+ */
+@Component(roles = DiagramMacroRunnable.class)
+@Singleton
+public class DiagramMacroRunnable extends AbstractDiagramRunnable
+{
+    private static final String MACRO_REFERENCE_PARAMETER = "reference";
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    @Named("compact")
+    private EntityReferenceSerializer<String> compactEntityReferenceSerializer;
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * @see com.xpn.xwiki.util.AbstractXWikiRunnable#runInternal()
+     */
+    @Override
+    public void runInternal()
+    {
+        while (!Thread.interrupted()) {
+            DiagramQueueEntry queueEntry;
+            BlockingQueue<DiagramQueueEntry> diagramsQueue = this.getDiagramsQueue();
+
+            try {
+                queueEntry = diagramsQueue.take();
+            } catch (InterruptedException e) {
+                logger.warn("Diagrams update thread has been interrupted", e);
+                queueEntry = STOP_RUNNABLE_ENTRY;
+            }
+
+            if (queueEntry == STOP_RUNNABLE_ENTRY) {
+                diagramsQueue.clear();
+                break;
+            }
```

Comment:
The code above looks pretty generic and could be moved to ``AbstractDiagramRunnable``, which would call an abstract ``processDiagram(queueEntry)``. WDYT?

Reply:
Indeed, it's better to make it a generic method since the same code is also used in DiagramLinksRunnable